app/assets/javascripts/discourse/app/controllers/second-factor-auth.js

```diff
@@ -194,7 +194,11 @@ export default Controller.extend({
           type: response.callback_method,
           data: { second_factor_nonce: this.nonce },
         })
-          .then(() => DiscourseURL.routeTo(response.redirect_path))
+          .then((callbackResponse) => {
+            const redirectUrl =
+              callbackResponse.redirect_url || response.redirect_url;
```

Comment:
Any chance they may be different? In what cases would they both exist?

Reply:
They shouldn't both exist at the same time. `callbackResponse` should contain the `redirect_url` when it's not possible to know the redirect URL until the user actually confirms their 2FA like in this case. But if the redirect URL can be determined when the 2FA confirmation process starts, then it should be present in `response` (this is the response of `POST /session/2fa`). An example of where the redirect URL is known upfront is the grant admin action:

https://github.com/discourse/discourse/blob/4c0a9fd4c30a13b60f5deffc8b2ed1d83254f97a/lib/second_factor/actions/grant_admin.rb#L16-L20

(this PR renames `redirect_path` to `redirect_url`)

If they both exist then that the means the action class and/or the route that's protected with 2FA is misconfigured and there's bug that needs fixing.

Reply:
Very good explanation, thanks